### PR TITLE
[Inputmethod][TCSACR-212] Add missing APIs

### PR DESCRIPTION
**Pull request by Inhong**
### Description of Change ###
- Add missing APIs

### API Changes ###
 - ACR: TCSACR-212

### CODE REVIEW

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs [hunk 1]
```diff
@@ -2209,5 +2215,186 @@ public static void SendCustomGeometryUpdated()
...
+        /// </privilege>
+        /// <param name="text">The selected text.</param>
+        /// <exception cref="InvalidOperationException">
+        /// This can occur due to the following reasons:
+        /// 1) The application does not have the privilege to call this function.
+        /// 2) The IME main loop has not started yet.
+        /// 3) Invalid parameter.
+        /// </exception>
+        /// <since_tizen> 6 </since_tizen>
+        public static void GetSelectedText(out string text)
```
Comment (myroot) [inline src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]:
```suggestion
        public static string GetSelectedText()
```

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs [hunk 2]
```diff
@@ -2209,5 +2215,186 @@ public static void SendCustomGeometryUpdated()
...
+                throw InputMethodExceptionFactory.CreateException(error);
+            }
+            text = Marshal.PtrToStringAnsi(txt);
+        }
+
+        /// <summary>
+        /// Called to set the prediction hint string to deliver to the input panel.
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public static event EventHandler<PredictionHintSetEventArgs> PredictionHintSet
```
Comment (myroot) [inline src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]:
```suggestion
        public static event EventHandler<PredictionHintSetEventArgs> PredictionHintUpdated
```

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs [hunk 3]
```diff
@@ -2209,5 +2215,186 @@ public static void SendCustomGeometryUpdated()
...
+        {
+            add
+            {
+                lock (thisLock)
+                {
+                    _imePredictionHintSetDelegate = (IntPtr predictionHint, IntPtr userData) =>
```
Comment (myroot) [inline src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]: always overwrite _imePredictionHintSetDelegate when added a event handler

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs [hunk 4]
```diff
@@ -2209,5 +2215,186 @@ public static void SendCustomGeometryUpdated()
...
+                    _predictionHintSet -= value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Called to set the prediction hint key and result to deliver to the input panel.
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public static event EventHandler<PredictionHintDataSetEventArgs> PredictionHintDataSet
```
Comment (myroot) [inline src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]:
```suggestion
        public static event EventHandler<PredictionHintDataSetEventArgs> PredictionHintDataUpdated
```

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs [hunk 5]
```diff
@@ -2209,5 +2215,186 @@ public static void SendCustomGeometryUpdated()
...
+                    _predictionHintDataSet -= value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Called when an associated text input UI control requests the text entry to set the MIME type.
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public static event EventHandler<MimeTypeSetRequestEventArgs> MimeTypeSetRequest
```
Comment (myroot) [inline src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]:
```suggestion
        public static event EventHandler<MimeTypeSetRequestEventArgs> MimeTypeUpdateRequested
```

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/PredictionHintDataSetEventArgs.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        {
+            get;
+            internal set;
+        }
+
+        /// <summary>
+        /// The prediction hint result.
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public string Result
```
Comment (myroot) [inline src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/PredictionHintDataSetEventArgs.cs]: What is result?
Result can only exist if there is a preceding request.

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs (hunk 1 above)
Comment (Inhong) [reply src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]: Fixed.

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs (hunk 2 above)
Comment (Inhong) [reply src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]: Fixed.

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs [hunk 7]
```diff
@@ -2209,5 +2215,186 @@ public static void SendCustomGeometryUpdated()
...
+            {
+                lock (thisLock)
+                {
+                    _predictionHintSet -= value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Called to set the prediction hint key and result to deliver to the input panel.
```
Comment (myroot) [inline src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]: It is hard to understand. what is result?
For what purpose is this event used?

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs (hunk 3 above)
Comment (Inhong) [reply src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]: I didn't understand your comment.

Comment (Inhong) [reply src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]: Fixed.

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs (hunk 7 above)
Comment (Inhong) [reply src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]: The result means key value. I will change result to value.

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/MimeTypeSetRequestEventArgs.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        internal MimeTypeSetRequestEventArgs(string mimeType)
+        {
+            MimeType = mimeType;
+        }
+
+        /// <summary>
+        /// The MIME type of entry.
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public string MimeType
```
[the remark this answers is not in the document]
Comment (Inhong) [reply src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/MimeTypeSetRequestEventArgs.cs]: Fixed.

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/PredictionHintDataSetEventArgs.cs [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        {
+            Key = key;
+            Result = result;
+        }
+
+        /// <summary>
+        /// The prediction hint key.
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public string Key
```
[the remark this answers is not in the document]
Comment (Inhong) [reply src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/PredictionHintDataSetEventArgs.cs]: Fixed.

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/PredictionHintSetEventArgs.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        internal PredictionHintSetEventArgs(string predictionHint)
+        {
+            PredictionHint = predictionHint;
+        }
+
+        /// <summary>
+        /// The prediction hint to be set to the input panel
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public string PredictionHint
```
[the remark this answers is not in the document]
Comment (Inhong) [reply src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/PredictionHintSetEventArgs.cs]: Fixed.

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs (hunk 3 above)
Comment (myroot) [reply src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]: Not yet fixed

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/PredictionHintDataSetEventArgs.cs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+        /// The prediction hint key.
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public string Key { get; }
+
+        /// <summary>
+        /// The prediction hint value.
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public string KeyValue { get; }
```
Comment (myroot) [inline src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/PredictionHintDataSetEventArgs.cs]: How about `Value`?
```suggestion
        public string Value{ get; }
```
This naming is used in KeyValuePair 
https://docs.microsoft.com/ko-kr/dotnet/api/system.collections.generic.keyvaluepair-2?view=netframework-4.7.2#%EC%86%8D%EC%84%B1

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs [hunk 12]
```diff
@@ -2209,5 +2215,187 @@ public static void SendCustomGeometryUpdated()
...
+                    _predictionHintSet -= value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Called to set the prediction hint key and value to deliver to the input panel.
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public static event EventHandler<PredictionHintDataSetEventArgs> PredictionHintDataUpdated
```
Comment (myroot) [inline src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]:
```suggestion
        public static event EventHandler<PredictionHintDataUpdatedEventArgs> PredictionHintDataUpdated
```

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs [hunk 13]
```diff
@@ -2209,5 +2215,187 @@ public static void SendCustomGeometryUpdated()
...
+                throw InputMethodExceptionFactory.CreateException(error);
+            }
+            return Marshal.PtrToStringAnsi(txt);
+        }
+
+        /// <summary>
+        /// Called to set the prediction hint string to deliver to the input panel.
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public static event EventHandler<PredictionHintSetEventArgs> PredictionHintUpdated
```
Comment (myroot) [inline src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]:
```suggestion
        public static event EventHandler<PredictionHintUpdatedEventArgs> PredictionHintUpdated
```

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs [hunk 14]
```diff
@@ -2209,5 +2215,187 @@ public static void SendCustomGeometryUpdated()
...
+                    _predictionHintDataSet -= value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Called when an associated text input UI control requests the text entry to set the MIME type.
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public static event EventHandler<MimeTypeSetRequestEventArgs> MimeTypeUpdateRequested
```
Comment (myroot) [inline src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]:
```suggestion
        public static event EventHandler<MimeTypeUpdateRequestedEventArgs> MimeTypeUpdateRequested
```

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs (hunk 13 above)
Comment (Inhong) [reply src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]: Fixed.

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs (hunk 14 above)
Comment (Inhong) [reply src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]: Fixed.

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs (hunk 12 above)
Comment (Inhong) [reply src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]: Fixed.

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs [hunk 15]
```diff
@@ -2209,5 +2215,187 @@ public static void SendCustomGeometryUpdated()
...
+        /// Called when an associated text input UI control requests the text entry to set the MIME type.
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public static event EventHandler<MimeTypeUpdateRequestedEventArgs> MimeTypeUpdateRequested
+        {
+            add
+            {
+                lock (thisLock)
+                {
+                    _imeMimeTypeSetRequestDelegate = (IntPtr mimeType, IntPtr userData) =>
```
Comment (myroot) [inline src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]: Overwritten

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/PredictionHintDataUpdatedEventArgs.cs [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+        /// The prediction hint key.
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public string Key { get; }
+
+        /// <summary>
+        /// The prediction hint value.
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public string KeyValue { get; }
```
Comment (myroot) [inline src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/PredictionHintDataUpdatedEventArgs.cs]: How about `Value`?
```suggestion
        public string Value{ get; }
```
This naming is used in KeyValuePair 
https://docs.microsoft.com/ko-kr/dotnet/api/system.collections.generic.keyvaluepair-2?view=netframework-4.7.2#%EC%86%8D%EC%84%B1

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs (hunk 15 above)
Comment (Inhong) [reply src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/InputMethodEditor.cs]: Fixed.

File: src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/PredictionHintDataUpdatedEventArgs.cs (hunk 16 above)
Comment (Inhong) [reply src/Tizen.Uix.InputMethod/Tizen.Uix.InputMethod/PredictionHintDataUpdatedEventArgs.cs]: Fixed.

**Pull request merged**